File: server/asset_service.go
package main

import (
	"errors"
	"net/http"

	"github.com/etherealmachine/rpg.ai/server/models"
)

type AssetService struct {
	db *models.Queries
}

type ListAssetsRequest struct {
}

type ListAssetsResponse struct {
	Spritesheets []models.Spritesheet
	Tilemaps     []models.TilemapWithThumbnails
}

func (s *AssetService) ListAssets(r *http.Request, args *ListAssetsRequest, reply *ListAssetsResponse) error {
	u := currentUser(r)
	if u == nil {
		return errors.New("no authenticated user found")
	}
	spritesheets, err := s.db.ListSpritesheetsByOwnerID(r.Context(), u.ID)
	if err != nil {
		return err
	}
	for _, s := range spritesheets {
		reply.Spritesheets = append(reply.Spritesheets, models.Spritesheet{
			ID:          s.ID,
			OwnerID:     u.ID,
			Name:        s.Name,
			Description: s.Description,
			Hash:        s.Hash,
			CreatedAt:   s.CreatedAt,
		})
	}
	tilemaps, err := s.db.ListTilemapsByOwnerID(r.Context(), u.ID)
	if err != nil {
		return err
	}
	for _, t := range tilemaps {
		reply.Tilemaps = append(reply.Tilemaps, models.TilemapWithThumbnails{
			Tilemap: models.Tilemap{
				ID:          t.ID,
				OwnerID:     u.ID,
				Name:        t.Name,
				Description: t.Description,
				Hash:        t.Hash,
				CreatedAt:   t.CreatedAt,
			},
		})
	}
	return nil
}

type DeleteAssetRequest struct {
	ID int32
}

type DeleteAssetResponse struct {
}

func (s *AssetService) DeleteSpritesheet(r *http.Request, args *DeleteAssetRequest, reply *DeleteAssetResponse) error {
	u := currentUser(r)
	if u == nil {
		return errors.New("no authenticated user found")
	}
	return s.db.DeleteSpritesheetWithOwner(r.Context(), models.DeleteSpritesheetWithOwnerParams{ID: args.ID, OwnerID: u.ID})
}

func (s *AssetService) DeleteTilemap(r *http.Request, args *DeleteAssetRequest, reply *DeleteAssetResponse) error {
	u := currentUser(r)
	if u == nil {
		return errors.New("no authenticated user found")
	}
	return s.db.DeleteTilemapWithOwner(r.Context(), models.DeleteTilemapWithOwnerParams{ID: args.ID, OwnerID: u.ID})
}

type ListSpritesheetsForTilemapRequest struct {
	TilemapID int32
}

type ListSpritesheetsForTilemapResponse struct {
	References []models.ListSpritesheetsForTilemapRow
}

func (s *AssetService) ListReferences(r *http.Request, args *ListSpritesheetsForTilemapRequest, reply *ListSpritesheetsForTilemapResponse) error {
	var err error
	reply.References, err = s.db.ListSpritesheetsForTilemap(r.Context(), args.TilemapID)
	if err != nil {
		return err
	}
	return nil
}

File: src/AssetService.ts
import JSONRPCService from './JSONRPCService';

export interface DeleteAssetRequest {
    ID: number
}

export interface DeleteAssetResponse {
}

export interface DeleteAssetRequest {
    ID: number
}

export interface DeleteAssetResponse {
}

export interface ListAssetsRequest {
}

export interface ListAssetsResponse {
    Spritesheets: Spritesheet[] | null
    Tilemaps: FilledTilemap[] | null
}

export interface ListSpritesheetsForTilemapRequest {
    TilemapID: number
}

export interface ListSpritesheetsForTilemapResponse {
    References: ListSpritesheetsForTilemapRow[] | null
}

export interface UpdateSpritesheetParams {
    ID: number
    OwnerID: number
    Name: string
    Description: NullString
    Definition: string
    Image: string
}

export interface Empty {
}

export interface UpdateTilemapParams {
    ID: number
    OwnerID: number
    Name: string
    Description: NullString
    Definition: string
}

export interface Empty {
}

export interface FilledTilemap {
        
    ID: number
    OwnerID: number
    Name: string
    Description: NullString
    Definition: string
    Hash: string
    CreatedAt: Date

    Thumbnails: Thumbnail[] | null
}

export interface ListSpritesheetsForTilemapRow {
    TilemapID: number
    SpritesheetID: number
    SpritesheetName: string
    SpritesheetHash: string
    SpritesheetDefinition: string
}

export interface NullInt32 {
    Int32: number
    Valid: boolean
}

export interface NullString {
    String: string
    Valid: boolean
}

export interface Spritesheet {
    ID: number
    OwnerID: number
    Name: string
    Description: NullString
    Definition: string
    Image: string
    Hash: string
    CreatedAt: Date
}

export interface Thumbnail {
    ID: number
    TilemapID: NullInt32
    SpritesheetID: NullInt32
    Image: string
    Hash: string
    ContentType: string
    Width: number
    Height: number
    CreatedAt: Date
}

export interface Tilemap {
    ID: number
    OwnerID: number
    Name: string
    Description: NullString
    Definition: string
    Hash: string
    CreatedAt: Date
}

class AssetService extends JSONRPCService {
    async DeleteSpritesheet(args: DeleteAssetRequest): Promise<DeleteAssetResponse> {
      return this.jsonrpc<DeleteAssetResponse>("DeleteSpritesheet", args);
    }
    async DeleteTilemap(args: DeleteAssetRequest): Promise<DeleteAssetResponse> {
      return this.jsonrpc<DeleteAssetResponse>("DeleteTilemap", args);
    }
    async ListAssets(args: ListAssetsRequest): Promise<ListAssetsResponse> {
      return this.jsonrpc<ListAssetsResponse>("ListAssets", args);
    }
    async ListReferences(args: ListSpritesheetsForTilemapRequest): Promise<ListSpritesheetsForTilemapResponse> {
      return this.jsonrpc<ListSpritesheetsForTilemapResponse>("ListReferences", args);
    }
    async UpdateSpritesheet(args: UpdateSpritesheetParams): Promise<Empty> {
      return this.jsonrpc<Empty>("UpdateSpritesheet", args);
    }
    async UpdateTilemap(args: UpdateTilemapParams): Promise<Empty> {
      return this.jsonrpc<Empty>("UpdateTilemap", args);
    }
}

const service = new AssetService("AssetService");
(window as any).AssetService = service;
export default service;
